// @lls/react-light-calendar/dist/index.js
"use strict";

function e(e) {
    return e && "object" == typeof e && "default" in e ? e.default : e
}
var p = e(require("lodash.times")),
    y = e(require("timestamp-utils")),
    t = require("react"),
    f = e(t),
    a = require("prop-types"),
    i = function(e, t) {
        if (!(e instanceof t)) throw new TypeError("Cannot call a class as a function")
    },
    n = function() {
        function n(e, t) {
            for (var a = 0; a < t.length; a++) {
                var n = t[a];
                n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(e, n.key, n)
            }
        }
        return function(e, t, a) {
            return t && n(e.prototype, t), a && n(e, a), e
        }
    }(),
    h = Object.assign || function(e) {
        for (var t = 1; t < arguments.length; t++) {
            var a = arguments[t];
            for (var n in a) Object.prototype.hasOwnProperty.call(a, n) && (e[n] = a[n])
        }
        return e
    },
    r = function(e, t) {
        if ("function" != typeof t && null !== t) throw new TypeError("Super expression must either be null or a function, not " + typeof t);
        e.prototype = Object.create(t && t.prototype, {
            constructor: {
                value: e,
                enumerable: !1,
                writable: !0,
                configurable: !0
            }
        }), t && (Object.setPrototypeOf ? Object.setPrototypeOf(e, t) : e.__proto__ = t)
    },
    c = function(e, t) {
        if (!e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        return !t || "object" != typeof t && "function" != typeof t ? e : t
    },
    g = function(e, t) {
        if (Array.isArray(e)) return e;
        if (Symbol.iterator in Object(e)) return function(e, t) {
            var a = [],
                n = !0,
                r = !1,
                o = void 0;
            try {
                for (var s, l = e[Symbol.iterator](); !(n = (s = l.next()).done) && (a.push(s.value), !t || a.length !== t); n = !0);
            } catch (e) {
                r = !0, o = e
            } finally {
                try {
                    !n && l.return && l.return()
                } finally {
                    if (r) throw o
                }
            }
            return a
        }(e, t);
        throw new TypeError("Invalid attempt to destructure non-iterable instance")
    },
    u = [31, null, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31],
    s = function(e) {
        e = e || (new Date).getTime();
        var t, a = y.decompose(e),
            n = g(a, 3),
            r = n[0],
            o = n[1],
            s = D(y.addDays(e, 1 - n[2])),
            l = u[o - 1] || ((t = r) % 4 == 0 && t % 100 != 0 || t % 400 == 0 ? 29 : 28),
            i = y.addDays(s, l - 1),
            c = y.getWeekDay(s);
        return {
            firstMonthDay: s,
            lastMonthDay: i,
            firstDayToDisplay: y.addDays(s, -c),
            month: o,
            year: r
        }
    },
    l = function(e, t) {
        return {
            startDate: t ? Math.min(e, t) : e,
            endDate: t && t !== e ? Math.max(e, t) : null
        }
    },
    D = function(e) {
        var t = y.decompose(e),
            a = g(t, 7);
        return y.add(e, {
            hours: -a[3],
            minutes: -a[4],
            seconds: -a[5],
            milliseconds: -a[6]
        })
    },
    d = function(e) {
        return ("0" + e).slice(-2)
    },
    v = function(e) {
        function o() {
            var e, t, l;
            i(this, o);
            for (var a = arguments.length, n = Array(a), r = 0; r < a; r++) n[r] = arguments[r];
            return (t = l = c(this, (e = o.__proto__ || Object.getPrototypeOf(o)).call.apply(e, [this].concat(n)))).onHoursChange = function(e) {
                var t = l.props;
                (0, t.onTimeChange)(y.setHours(t.date, parseInt(e.target.value, 10)))
            }, l.onMinutesChange = function(e) {
                var t = l.props;
                (0, t.onTimeChange)(y.setMinutes(t.date, parseInt(e.target.value, 10)))
            }, l.render = function() {
                var e = l.props,
                    t = e.date,
                    a = e.displayTime,
                    n = e.dayLabels,
                    r = e.monthLabels,
                    o = y.getHours(t),
                    s = y.getMinutes(t);
                return f.createElement("div", {
                    className: "rlc-date-details-wrapper"
                }, f.createElement("div", {
                    className: "rlc-date-details"
                }, f.createElement("div", {
                    className: "rlc-date-number"
                }, y.getDay(t)), f.createElement("div", {
                    className: "rlc-date-day-month-year"
                }, f.createElement("div", {
                    className: "rlc-detail-day"
                }, n[y.getWeekDay(t)]), f.createElement("div", {
                    className: "rlc-detail-month-year"
                }, r[y.getMonth(t) - 1], " ", f.createElement("span", {
                    className: "rlc-detail-year"
                }, y.getYear(t))))), a && f.createElement("div", {
                    className: "rlc-date-time-selects"
                }, f.createElement("select", {
                    onChange: l.onHoursChange,
                    value: o
                }, p(24).map(function(e) {
                    return f.createElement("option", {
                        value: d(e),
                        key: e
                    }, d(e))
                })), f.createElement("span", {
                    className: "rlc-time-separator"
                }, ":"), f.createElement("select", {
                    onChange: l.onMinutesChange,
                    value: s
                }, p(60).map(function(e) {
                    return f.createElement("option", {
                        value: d(e),
                        key: e
                    }, d(e))
                }))))
            }, c(l, t)
        }
        return r(o, t.Component), o
    }();
v.propTypes = {
    date: a.number,
    displayTime: a.bool,
    dayLabels: a.arrayOf(a.string),
    monthLabels: a.arrayOf(a.string),
    onTimeChange: a.func
};
var C = function(e) {
    function s() {
        var e, t, r;
        i(this, s);
        for (var a = arguments.length, n = Array(a), o = 0; o < a; o++) n[o] = arguments[o];
        return (t = r = c(this, (e = s.__proto__ || Object.getPrototypeOf(s)).call.apply(e, [this].concat(n)))).prevYear = function() {
            return r.props.onChange({
                yearOffset: -1
            })
        }, r.prevMonth = function() {
            return r.props.onChange({
                monthOffset: -1
            })
        }, r.nextYear = function() {
            return r.props.onChange({
                yearOffset: 1
            })
        }, r.nextMonth = function() {
            return r.props.onChange({
                monthOffset: 1
            })
        }, r.render = function() {
            var e = r.props,
                t = e.monthLabels,
                a = e.month,
                n = e.year;
            return f.createElement("div", {
                className: "rlc-month-and-year-wrapper"
            }, f.createElement("div", {
                className: "rlc-navigation-button-wrapper rlc-prevs"
            }, f.createElement("div", {
                className: "rlc-navigation-button rlc-prev-year",
                onClick: r.prevYear
            }, "<<"), f.createElement("div", {
                className: "rlc-navigation-button rlc-prev-month",
                onClick: r.prevMonth
            }, "<")), f.createElement("div", {
                className: "rlc-month-and-year"
            }, t[a - 1], " ", f.createElement("span", null, n)), f.createElement("div", {
                className: "rlc-navigation-button-wrapper rlc-nexts"
            }, f.createElement("div", {
                className: "rlc-navigation-button rlc-next-month",
                onClick: r.nextMonth
            }, ">"), f.createElement("div", {
                className: "rlc-navigation-button rlc-next-year",
                onClick: r.nextYear
            }, ">>")))
        }, c(r, t)
    }
    return r(s, t.Component), s
}();
C.propTypes = {
    monthLabels: a.arrayOf(a.string),
    month: a.string,
    year: a.string,
    onChange: a.func
};
var o = function(e) {
    function o() {
        var e, t, b;
        i(this, o);
        for (var a = arguments.length, n = Array(a), r = 0; r < a; r++) n[r] = arguments[r];
        return (t = b = c(this, (e = o.__proto__ || Object.getPrototypeOf(o)).call.apply(e, [this].concat(n)))).state = {}, b.onClickDay = function(e) {
            var t = b.state,
                a = t.startDate;
            b.update(a ? a && !t.endDate ? l(a, e) : {
                startDate: e,
                endDate: null
            } : {
                startDate: e
            })
        }, b.changeMonth = function(e) {
            var t = e.yearOffset,
                a = e.monthOffset,
                n = y.add(b.state.firstMonthDay, {
                    months: void 0 === a ? 0 : a,
                    years: void 0 === t ? 0 : t
                });
            b.setState(s(n))
        }, b.update = function(e) {
            var t = e.startDate,
                a = e.endDate;
            b.props.onChange(void 0 === t ? b.props.startDate : t, void 0 === a ? b.props.endDate : a)
        }, b.getClassNames = function(e) {
            var t, a, n, r, o, s, l, i, c, u = b.state,
                d = u.firstMonthDay,
                m = u.lastMonthDay,
                p = u.endDate,
                y = b.props.disableDates,
                f = D(u.startDate),
                h = D(p),
                v = (t = {
                    "rlc-day-disabled": y(e),
                    "rlc-day-today": e === D((new Date).getTime()),
                    "rlc-day-inside-selection": (l = e, i = f, c = h, i < l && l < c),
                    "rlc-day-out-of-month": (r = e, o = d, s = m, r < o || s < r),
                    "rlc-day-selected": !p && f === e,
                    "rlc-day-start-selection": p && f === e,
                    "rlc-day-end-selection": p && h === e
                }, n = !0, (a = "rlc-day-" + e) in t ? Object.defineProperty(t, a, {
                    value: n,
                    enumerable: !0,
                    configurable: !0,
                    writable: !0
                }) : t[a] = n, t);
            return Object.entries(v).reduce(function(e, t) {
                var a = g(t, 2);
                return a[1] ? e + " " + a[0] : e
            }, "") || "Day_default"
        }, b.render = function() {
            var t, e = b.state,
                a = e.firstDayToDisplay,
                n = e.startDate,
                r = e.endDate,
                o = e.month,
                s = e.year,
                l = b.props,
                i = l.disableDates,
                c = l.displayTime,
                u = l.dayLabels,
                d = l.monthLabels,
                m = function(e, t) {
                    var a = {};
                    for (var n in e) 0 <= t.indexOf(n) || Object.prototype.hasOwnProperty.call(e, n) && (a[n] = e[n]);
                    return a
                }(l, ["startDate", "endDate", "onChange", "disableDates", "displayTime", "dayLabels", "monthLabels", "timezone"]);
            return f.createElement("div", h({
                className: "rlc-calendar"
            }, m), f.createElement("div", {
                className: "rlc-details"
            }, n && f.createElement(v, {
                dayLabels: u,
                monthLabels: d,
                date: n,
                displayTime: c,
                onTimeChange: function(e) {
                    return b.update({
                        startDate: e
                    })
                }
            }), r && f.createElement(v, {
                dayLabels: u,
                monthLabels: d,
                date: r,
                displayTime: c,
                onTimeChange: function(e) {
                    return b.update({
                        endDate: e
                    })
                }
            })), f.createElement(C, {
                monthLabels: d,
                month: o,
                year: s,
                onChange: b.changeMonth
            }), f.createElement("div", {
                className: "rlc-days-label"
            }, u.map(function(e) {
                return f.createElement("div", {
                    className: "rlc-day-label",
                    key: e.toLowerCase()
                }, e.slice(0, 2))
            })), f.createElement("div", {
                className: "rlc-days"
            }, (t = a, p(42, function(e) {
                return y.addDays(t, e)
            })).map(function(e) {
                return f.createElement("div", {
                    className: "rlc-day " + b.getClassNames(e),
                    key: e,
                    onClick: function() {
                        return !i(e) && b.onClickDay(e)
                    }
                }, parseInt(y.getDay(e), 10))
            })))
        }, c(b, t)
    }
    return r(o, t.PureComponent), n(o, null, [{
        key: "getDerivedStateFromProps",
        value: function(e) {
            var t = e.startDate,
                a = e.endDate;
            return y.setTimezone(e.timezone), h({}, s(t), l(t, a))
        }
    }]), o
}();
o.defaultProps = {
    startDate: null,
    endDate: null,
    onChange: function() {},
    disableDates: function() {
        return !1
    },
    displayTime: !1,
    dayLabels: ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"],
    monthLabels: ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"],
    timezone: "UTC"
}, o.propTypes = {
    startDate: a.number,
    endDate: a.number,
    onChange: a.func,
    disableDates: a.func,
    displayTime: a.bool,
    dayLabels: a.arrayOf(a.string),
    monthLabels: a.arrayOf(a.string),
    timezone: a.string
}, module.exports = o;
